### PR TITLE
Support webhooks for more applications types

### PR DESCRIPTION
**Pull request by alextargov**
**Description**

Changes proposed in this pull request:
- Support automatic webhook creation for two application types
- Support automatic webhook creation for the system fetcher flow, where we do TrustedUpserts
- Support automatic webhook creation for the manually registered systems. When a manual pairing is done, an external service calls our UpdateApplication mutation to set the baseURL. If such a request is done, we consider that the application may have been successfully paired and we create the webhook automatically.
- The systems come from the system fetcher component with a ppmsProductVersionId property. Some systems do not support ORD and we should not create webhooks for them. This is why there are guards in the System Fetcher flow that check this property and the supported values from our configuration
-  
**Related issue(s)**

- ...

**Pull Request status**

- [x] Implementation
- [x] Unit tests
- [x] Integration tests
- [x] `chart/compass/values.yaml` is updated 
- [x] Mocks are regenerated, using the automated script

### CODE REVIEW

File: components/director/internal/domain/application/service.go [hunk 1]
```diff
@@ -1356,6 +1430,85 @@ func (s *service) unassignFormations(ctx context.Context, appTenant, objectID st
 	return nil
 }
 
+func (s *service) getMappingORDConfiguration(applicationType string) (ORDWebhookMapping, bool) {
+	for _, wm := range s.ordWebhookMapping {
+		if wm.Type == applicationType {
+			return wm, true
+		}
+	}
+	return ORDWebhookMapping{}, false
```
Comment (dragobt) [inline components/director/internal/domain/application/service.go:1439]: maybe add unit test for this case

File: components/director/internal/domain/application/service.go [hunk 2]
```diff
@@ -1356,6 +1430,85 @@ func (s *service) unassignFormations(ctx context.Context, appTenant, objectID st
...
+
+func (s *service) prepareORDWebhook(ctx context.Context, baseURL, applicationType, ppmsProductVersionID string) *model.WebhookInput {
+	if baseURL == "" {
+		log.C(ctx).Infof("No baseURL found in input. Will not create a webhook")
+		return nil
+	}
+
+	mappingCfg, ok := s.getMappingORDConfiguration(applicationType)
+	if !ok {
+		log.C(ctx).Infof("Missing ord configuration for application type %q", applicationType)
```
Comment (dragobt) [inline components/director/internal/domain/application/service.go:1450]: same here

File: components/director/internal/domain/application/service.go [hunk 3]
```diff
@@ -1356,6 +1430,85 @@ func (s *service) unassignFormations(ctx context.Context, appTenant, objectID st
...
+func buildWebhookURL(suffix string, ordPath string, baseURL *string) (string, error) {
+	url, err := url.Parse(*baseURL)
+	if err != nil {
+		return "", err
+	}
+
+	hostParts := strings.Split(url.Host, urlSubdomainSeparator)
+
+	if !strings.HasSuffix(hostParts[0], suffix) {
+		hostParts[0] = fmt.Sprintf("%s%s", hostParts[0], suffix)
```
Comment (dragobt) [inline components/director/internal/domain/application/service.go:1487]: maybe unit test here as well

File: components/director/internal/domain/application/service.go [hunk 4]
```diff
@@ -524,6 +533,44 @@ func (s *service) Update(ctx context.Context, id string, in model.ApplicationUpd
...
+	if ordWebhook == nil {
+		log.C(ctx).Infof("Skipping ORD Webhook creation for app with id %q.", app.ID)
+		return nil
+	}
+
+	ordWebhooks := make([]*model.WebhookInput, 0)
+	ordWebhooks = append(ordWebhooks, ordWebhook)
+
+	if err = s.createWebhooksIfNotExist(ctx, app.ID, appTenant, ordWebhooks); err != nil {
+		return errors.Wrapf(err, "while processing webhooks for application with id %q", app.ID)
```
Comment (dragobt) [inline components/director/internal/domain/application/service.go:571]: unit test here

File: tests/director/tests/application_api_test.go [hunk 5]
```diff
@@ -431,47 +431,139 @@ func TestCreateApplicationWithNonExistentIntegrationSystem(t *testing.T) {
 }
 
 func TestUpdateApplication(t *testing.T) {
```
Comment (dragobt) [inline tests/director/tests/application_api_test.go:433]: do we need e2e test for the case where auto-discovered S/4 systems have `ppmsProductVersionId` which is less than the `ppmsProductVersionId` corresponding to S/4 HANA Cloud 2208 -> so we should not create ORD webhook? What do you think?

File: components/director/internal/domain/application/service.go [hunk 6]
```diff
@@ -524,6 +533,44 @@ func (s *service) Update(ctx context.Context, id string, in model.ApplicationUpd
...
+
+	ppmsProductVersionIDLbl, err := s.labelService.GetByKey(ctx, appTenant, model.ApplicationLabelableObject, app.ID, ppmsProductVersionIDLabelKey)
+	if err != nil {
+		if !apperrors.IsNotFoundError(err) {
+			return errors.Wrapf(err, "while getting label %q for %q with id %q", ppmsProductVersionIDLabelKey, model.ApplicationLabelableObject, app.ID)
+		}
+	}
+
+	ppmsProductVersionID := ""
+	if ppmsProductVersionIDLbl != nil {
```
Comment (kaloyanspiridonov) [inline components/director/internal/domain/application/service.go:555]: && ppmsProductVersionIDLbl.Value != nil

File: components/director/internal/domain/application/service.go [hunk 7]
```diff
@@ -524,6 +533,44 @@ func (s *service) Update(ctx context.Context, id string, in model.ApplicationUpd
...
+	}
+
+	ordWebhook := s.prepareORDWebhook(ctx, str.PtrStrToStr(in.BaseURL), appTypeLbl.Value.(string), ppmsProductVersionID)
+	if ordWebhook == nil {
+		log.C(ctx).Infof("Skipping ORD Webhook creation for app with id %q.", app.ID)
+		return nil
+	}
+
+	ordWebhooks := make([]*model.WebhookInput, 0)
+	ordWebhooks = append(ordWebhooks, ordWebhook)
```
Comment (kaloyanspiridonov) [inline components/director/internal/domain/application/service.go:568]: [1]*model.WebhookInput{ordWebhook}

File: components/director/internal/domain/application/config.go [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+// ORDWebhookMapping represents a struct for ORD Webhook Mappings
+type ORDWebhookMapping struct {
+	Type                string   `json:"Type"`
+	PpmsProductVersions []string `json:"PpmsProductVersions"`
+	OrdURLPath          string   `json:"OrdUrlPath"`
+	SubdomainSuffix     string   `json:"SubdomainSuffix"`
+}
+
+// UnmarshalMappings unmarshalls a string into []ORDWebhookMapping. This is done because of limitation of the envconfig library
+func UnmarshalMappings(mappingsConfig string) ([]ORDWebhookMapping, error) {
```
[the remark this answers is not in the document]
Comment (kaloyanspiridonov) [reply components/director/internal/domain/application/config.go:16]: testing unmarshal func for me is unnecessary

File: tests/director/tests/application_api_test.go (hunk 5 above)
Comment (alextargov) [reply tests/director/tests/application_api_test.go:433]: I think we have covered that in the unit tests, so probably threre's no need for an e2e test

**Review by dragobt — APPROVED** [review]

**Pull request merged**